src/window.rs

```diff
@@ -1396,6 +1396,15 @@ impl Window {
             .maybe_wait_on_main(|w| w.drag_resize_window(direction))
     }
 
+    /// Show system menu at current mouse position.
+    ///
+    /// ## Platform-specific
+    ///
+    /// **Android / iOS / macOS / Orbital / Wayland / Web / X11:** Unsupported.
+    pub fn show_system_menu(&self) {
```

Comment:
I think we shouldn't use mouse position for this, it seems like we can use arbitrary position based on your implementation, so just passing `Position` as argument should be good.

That way we can remove the mouse based wording, and say that we request to show window menu at the given `position`.

Also, could we name it `show_window_menu` and also describe what it is and how it usually looks like and document the use of?

Reply:
> Given that there was a mention of the `mouse` or similar happening, I'd say that wayland wants that action to be associated with some input event, like in reaction to mouse click, keyboard button press, touch press, etc.

> I think we shouldn't use mouse position for this, it seems like we can use arbitrary position based on your implementation, so just passing `Position` as argument should be good.
> 
> That way we can remove the mouse based wording, and say that we request to show window menu at the given `position`.

Just to clarify: if we add a position parameter we don't need to associate it with some input event, right?

> We should say that the compositor is free to ignore such request though.

It would be nice if we can figure out the details here. When the compositor ignores this and platform-specific behavior if it differs.

> Also, could we name it `show_window_menu` [..]

Is there actually some defined term for this? I found the term "system menu" for Windows, do Wayland or X11 have their own terms?

See https://github.com/rust-windowing/winit/pull/3109#discussion_r1334347502 for previous discussion.

Reply:
@daxpedda I mean, it's still based on the event, as in, you have to do it, but the position could be arbitrary and not attached to anything. But that's what wayland wants for that, because you have a concept of serial events, so you just must pass a serial from the latest button event.

>It would be nice if we can figure out the details here. When the compositor ignores this and platform-specific behavior if it differs.

it's a compositor menu in the first place, I'm pretty sure no-one ever guarantees that they will show the menu in the end, and it's transparent for the window/winit (at least on Wayland). I'm sure it's similar on windows, since I don't think user is the one closing it, it's done for him.

>Is there actually some defined term for this? I found the term "system menu" for Windows, do Wayland or X11 have their own terms?

https://wayland.app/protocols/xdg-shell#xdg_toplevel:request:show_window_menu 

That's a wayland API for that.

Reply:
> @daxpedda I mean, it's still based on the event, as in, you have to do it, but the position could be arbitrary and not attached to anything. But that's what wayland wants for that, because you have a concept of serial events, so you just must pass a serial from the latest button event.

If I understand that correctly, you are saying that you can pass an arbitrary position, but you need the serial from the last button event. Does it matter how long ago this button even was as long as it's the last one? Is this the case for windows as well @YouKnow-sys?

> it's a compositor menu in the first place, I'm pretty sure no-one ever guarantees that they will show the menu in the end, and it's transparent for the window/winit (at least on Wayland). I'm sure it's similar on windows, since I don't think user is the one closing it, it's done for him.

I'm not entirely sure if we can equate closing with not showing, but maybe you mean something else? It would be nice if can actually document this in clear terms, like **in which case** does it not show, is it random? @YouKnow-sys is this the case for Windows as well?

If we just don't know the answer to this we can just write "no guarantee" as @YouKnow-sys had before.

Reply:
>Does it matter how long ago this button even was as long as it's the last one? Is this the case for windows as well

Compositor defined, not specified, just not long ago.

@daxpedda I mean both, system compositor is in-charge of the menu, it can either not show at all or close without user input.

No guarantee means that it simply may not be showed to the user and winit has no control over such menu.

Wayland docs say the following for example:

>There are no guarantees as to what menu items the window menu contains, or even if a window menu will be drawn at all.

Reply:
Alright then. Would be nice to confirm exactly whats going on in Windows to figure out if this is platform-specific or not.

Reply:
I don't think this menu thing differs that much tbh, it would be nice to confirm, but just so you know, it's not a **new** thing, we already show it iirc. It just we don't have a way to show it when there're no decorations at all, but default system decorations usually embed such a thing.

Reply:
I don't think we should mention it's not guaranteed if that's not the case for all platforms. It should go into platform-specifics in that case, which isn't necessary right now because we only have the Windows implementation.

Winit doesn't offer a way to remove `WS_SYSMENU`, right? Because then I don't think we should mention it at all, it becomes a internal detail at this point.

Thank you for verifying this @YouKnow-sys! I will make some suggestions so you don't have to guess around what we want here!

Reply:
About the mouse position: personally I don't particularly see the point of supplying one because I assume that implementations will only want to show this menu at the mouse position anyway. So I'm fine with how it is right now, but I'm not against adding an optional position.

I think **requiring** a position is problematic in Winit right now because we are missing https://github.com/rust-windowing/winit/issues/883 or https://github.com/rust-windowing/winit/pull/2648, which would require users to save the last known position somewhere. Not that this is a problem, I just don't know what the motivation would be to do this in the first place.

Reply:
> Although we can still get the window handle and modify the flags ourselves using win32 API, so maybe it does worth having in doc (so user don't accidentally remove this flag)

I'm not familiar with the Windows backend, but couldn't adding or removing flags break a whole lot anyway? At that point I would argue we *maybe* want to add some documentation where you can get the handle, telling users that with this handle they can do all sort of shenanigans and break Winit completely.

Reply:
> although Im not sure what should the position type be... should it be a tuple of `u16` `(u16, u16)`?

`Into<Position>`, like in [`Window::set_cursor_position()`](https://docs.rs/winit/0.29.1-beta/winit/window/struct.Window.html#method.set_cursor_position). It might not work with an `Option` though, in which case just use [`Position`](https://docs.rs/winit/0.29.1-beta/winit/dpi/enum.Position.html).

Reply:
> `TrackPopupMenu` in win32 expect the position to be in screen coordinates, how can exactly get screen coordinates from `Position`?

You will have to do some kind of calculation. Just looking at [`Window::set_cursor_position()`](https://github.com/rust-windowing/winit/blob/2422ea39d0e97fd43698391f84d9300cd169d8cd/src/platform_impl/windows/window.rs#L415-L418), it's using [`ClientToScreen`](https://docs.rs/windows-sys/0.48.0/windows_sys/Win32/Graphics/Gdi/fn.ClientToScreen.html), which according to the [docs](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-clienttoscreen) seems to be what you are looking for.

> also the thing about using `Into<Position>` is if we use just None now we have to use type annotations as well, because then rust wont know what the type is... is there any better option?

> It might not work with an `Option` though, in which case just use [`Position`](https://docs.rs/winit/0.29.1-beta/winit/dpi/enum.Position.html).

That's what I meant here, just use `Option<Position>` instead of `Option<impl Into<Position>>`.